### PR TITLE
Add optional token for regex and httpheader. Add htmlparser source

### PR DESCRIPTION
**Pull request by ypsilik**
For application with internal access, we can use authorization token to access the url.

- regex source has an optional token option available sent in the url headers
- httpheader have an optional token option available sent in the url headers

Another use case is parsing html output with xpath to get the release.

- A source "htmlparser" have been added that allow to use xpath to get the release. It also take optionnal token.

I have update docs, files and I try to make new tests.
I'm open for any reviews and comment, to get a better code.
Thanks

### CODE REVIEW

File: nvchecker_source/httpheader.py [hunk 1]
```diff
@@ -3,37 +3,53 @@
...
+    )
+
+    header_value = res.headers.get(header)
+    if not header_value:
+        raise GetVersionError('header %s not found or is empty' % header)
+    try:
+        version = regex.findall(header_value)
+    except ValueError:
+        raise GetVersionError('version string not found.')
+    return version
```
Comment (lilydjwg) [inline nvchecker_source/httpheader.py:55]: Please don't reformat existing files you don't intend to change.

**Review by lilydjwg — COMMENTED** [review]
Since we're adding the `token` option to a lot of HTTP requests, why don't we add to all of them like `proxy` and `user_agent` options?

File: nvchecker_source/httpheader.py (hunk 1 above)
Comment (ypsilik) [reply nvchecker_source/httpheader.py:55]: Sorry, my editor autoformats the files, I will be more careful.

**Comment by ypsilik** [comment]
I think this can be a solution.
Currently the tokens in the key file have different names depending on the http source chosen. Is this not a problem if the option is global?

**Comment by lilydjwg** [comment]
I think there is no need to distinguish these tokens used by different sources. The `name` is unique so something like `httptoken_{name}` should suffice.

**Comment by ypsilik** [comment]
Yes, you are right, thanks for the review, I am working on it to update it.

**Comment by ypsilik** [comment]
Hello, I have updated the file. I added the httptoken in the global settings. I am not sure about the ctxvars settings.

File: nvchecker_source/htmlparser.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+async def get_version(name, conf, *, cache, **kwargs):
+  key = tuple(sorted(conf.items()))
+  return await cache.get(key, get_version_impl)
+
+async def get_version_impl(info):
+  conf = dict(info)
+  encoding = conf.get('encoding', 'latin1')
+
+  res = await session.get(conf['url'])
+  body = html.fromstring(res.body.decode(encoding))
```
Comment (lilydjwg) [inline nvchecker_source/htmlparser.py:20]: lxml can parse the encoding declaration from the HTML, so we can delegate that to lxml if not specified.

```py
  encoding = conf.get('encoding')
  parser = html.HTMLParser(encoding=encoding)
  doc = html.fromstring(res.body, base_url=r.url, parser=parser)
```

It's possible to get the text encoding from `Content-Type` header, but maybe it's too much effort to parse that.

File: nvchecker_source/htmlparser.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  try:
+    checkxpath = body.xpath(conf.get('xpath'))
+  except etree.XPathEvalError as e:
+    raise GetVersionError('bad xpath', exc_info=e)
```
Comment (lilydjwg) [inline nvchecker_source/htmlparser.py:24]: Why do we need to check first? We should be able to catch the exception if it's bad.

**Review by lilydjwg — CHANGES_REQUESTED** [review]

**Comment by lilydjwg** [comment]
I haven't tested yet, but the `httptoken` thing should be alright.

File: nvchecker_source/htmlparser.py (hunk 2 above)
Comment (ypsilik) [reply nvchecker_source/htmlparser.py:20]: You're right, I changed that.

Comment (lilydjwg) [reply nvchecker_source/htmlparser.py:20]: Did you forget to push your commits? I do not see any new commits yet.

Comment (ypsilik) [reply nvchecker_source/htmlparser.py]: indeed, it's done.

File: docs/usage.rst [hunk 4]
```diff
@@ -262,6 +270,19 @@ method
...
+
+  source = "htmlparser"
+
+Send an HTTP request and search through the body a specific xpath.
+
+url
+  The URL of the HTTP request.
+
+xpath
+  A xpath expression used to find the version string
```
Comment (lilydjwg) [inline docs/usage.rst:285]:
```suggestion
  An xpath expression used to find the version string.
```

File: docs/usage.rst [hunk 5]
```diff
@@ -176,6 +176,14 @@ tries
 
   This only works when the source implementation uses the builtin HTTP client.
 
+httptoken
+  A personal authorization token used to call the url. The type of token depends on the authorization required.
```
Comment (lilydjwg) [inline docs/usage.rst:180]:
```suggestion
  A personal authorization token used to fetch the url with the ``Authorization`` header. The type of token depends on the authorization required. 
```

File: nvchecker/util.py [hunk 6]
```diff
@@ -224,7 +225,13 @@ async def run_one(
...
     if ua is not None:
       ctx_ua.set(ua)
-
+    httpt = entry.get('httptoken', None)
+    if httpt is not None:
+      ctx_httpt.set(httpt)
+    else:
+      httpt = self.keymanager.get_key('httptoken_'+name)
+      if httpt is not None:
+        ctx_httpt.set(httpt)
```
Comment (lilydjwg) [inline nvchecker/util.py:234]:
```suggestion
    if httpt is None:
      httpt = self.keymanager.get_key('httptoken_'+name)
    if httpt is not None:
        ctx_httpt.set(httpt)
```

File: nvchecker_source/htmlparser.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# MIT licensed
+# Copyright (c) 2020 Ypsilik <tt2laurent.maud@gmail.com>, et al.
+# Copyright (c) 2013-2020 lilydjwg <lilydjwg@gmail.com>, et al.
+
+import re
+import sre_constants
```
Comment (lilydjwg) [inline nvchecker_source/htmlparser.py:6]: Why is this `sre_constants` imported?

File: docs/usage.rst [hunk 8]
```diff
@@ -262,6 +270,19 @@ method
 follow_redirects
   (*Optional*) Whether to follow 3xx HTTP redirects. Default is ``false``. If you are looking at a ``Location`` header, you shouldn't change this.
 
+Search with an HTML Parser
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+::
+
+  source = "htmlparser"
+
+Send an HTTP request and search through the body a specific xpath.
```
Comment (lilydjwg) [inline docs/usage.rst:279]: We need to note that an additional dependency `lxml` is required.

**Review by lilydjwg — CHANGES_REQUESTED** [review]
Sorry for the delay. I was busy and this patch is really large to look at.

File: docs/usage.rst (hunk 8 above)
Comment (ypsilik) [reply docs/usage.rst:279]: Should I add the dependency in the setup.py "install_requires" ?

Comment (lilydjwg) [reply docs/usage.rst:279]: No, because only users who use this source plugin need it, and lxml is quite large.

**Pull request merged**